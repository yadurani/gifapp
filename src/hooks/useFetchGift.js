import { useState, useEffect } from 'react';

import getGif from '../helpers/getGift';

const useFetchGift = (category) => {
  const [offset, setOffset] = useState(0)
  const [state, setstate] = useState({
    data: [],
    pagination: null,
    loading: true,
  })
  const [disableBtn, setDisabledBtn] = useState(false)

  const fetchImages = async () => {
    const response = await getGif(category, offset * 10)
    setstate(prevState => ({
      ...prevState,
      data: [ ...response.gifs, ...prevState.data ],
      pagination: response.pagination.total_count,
      offset: offset * 10,
      loading: false,
    }))
  }

  const handlePage = () => {
    if(state.offset >= state.pagination) {
      setDisabledBtn(true)
    }
    if(state.offset <= state.pagination){
      setOffset(prevState => prevState + 1)
    }
  }
  
  useEffect(() => {
    fetchImages()
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [category, offset])

  return {state, disableBtn, handlePage}
}

export default useFetchGift
