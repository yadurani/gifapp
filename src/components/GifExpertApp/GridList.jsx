import React from 'react';
import PropTypes from 'prop-types';

import getDataNoDuplicated from '../../helpers/noDuplicated';
import useFetchGift from '../../hooks/useFetchGift';

import GridItem from './GridItem';
import LoadingSkeleton from './LoadingSkeleton';

const GridList = ({ category }) => {
  const { state, handlePage, disableBtn } = useFetchGift(category)
  const { data: images, loading  } = state
  const imagesFinally = getDataNoDuplicated(images)
  const noData = imagesFinally.length === 0

  return (
    <>
      <div className="subtitle">
        <h3 className="animate__animated animate__fadeIn">{category}</h3>
        <hr/>
        {!noData && <button disabled={disableBtn} onClick={handlePage}>Ver m√°s</button>}
      </div>
      {noData && !loading && <div className="notfound">Ups!! Not found, try again.</div>}
      <div className="card-grid">
        {loading && <LoadingSkeleton/>}
        {imagesFinally?.map(({ id, ...rest }) => <GridItem key={id} {...rest}/>)}
      </div>
    </>
  )
};

GridList.propTypes = {
  category: PropTypes.string.isRequired
}

export default GridList
