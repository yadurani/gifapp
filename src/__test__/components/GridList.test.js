import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'

import GridList from '../../components/GifExpertApp/GridList'
import useFetchGift from '../../hooks/useFetchGift'
jest.mock('../../hooks/useFetchGift')

describe('Probando el componente <GridList />', () => {
  const category = 'One punch';
  
  it('Debe hacer el toMatchSnapshot, componente cargando', () => {
    useFetchGift.mockReturnValue({
      state: {
        data: [],
        loading: true,
      }
    })
    
    const wrapper = shallow(<GridList category={category}/>);
    expect(wrapper).toMatchSnapshot();
  })

  it('Debe hacer el toMatchSnapshot, componente not found', () => {
    useFetchGift.mockReturnValue({
      state: {
        data: [],
        loading: false,
      }
    })
    
    const wrapper = shallow(<GridList category={category}/>);
    expect(wrapper).toMatchSnapshot();
  })

  it('Debe de mostrar items cuando se cargan imagÃ©nes useFetchGifs', () => {
    const gifs = [{
      id: 'ABC',
      url: 'https://localhost:3000/cosa.jpg',
      title: 'Cualquier cosa'
    }]
    useFetchGift.mockReturnValue({
      state: {
        data: gifs,
        loading: false,
      }
    })

    const wrapper = shallow(<GridList category={category}/>);
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('LoadingSkeleton').exists()).toBe(false)
    expect(wrapper.find('.notfound').exists()).toBe(false)
    expect(wrapper.find('GridItem').length).toBe(gifs.length)
  })
})