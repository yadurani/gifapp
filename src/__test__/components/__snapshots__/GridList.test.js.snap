// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente <GridList /> Debe de mostrar items cuando se cargan imagénes useFetchGifs 1`] = `
<Fragment>
  <div
    className="subtitle"
  >
    <h3
      className="animate__animated animate__fadeIn"
    >
      One punch
    </h3>
    <hr />
    <button>
      Ver más
    </button>
  </div>
  <div
    className="card-grid"
  >
    <GridItem
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost:3000/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Probando el componente <GridList /> Debe hacer el toMatchSnapshot, componente cargando 1`] = `
<Fragment>
  <div
    className="subtitle"
  >
    <h3
      className="animate__animated animate__fadeIn"
    >
      One punch
    </h3>
    <hr />
  </div>
  <div
    className="card-grid"
  >
    <LoadingSkeleton />
  </div>
</Fragment>
`;

exports[`Probando el componente <GridList /> Debe hacer el toMatchSnapshot, componente not found 1`] = `
<Fragment>
  <div
    className="subtitle"
  >
    <h3
      className="animate__animated animate__fadeIn"
    >
      One punch
    </h3>
    <hr />
  </div>
  <div
    className="notfound"
  >
    Ups!! Not found, try again.
  </div>
  <div
    className="card-grid"
  />
</Fragment>
`;
